File: gcp-api/src/lib/cloud_vision_validator.py
"""
Google Cloud Vision API-based photo validation module.
"""

from __future__ import annotations

import logging
import cv2
import numpy as np
from enum import Enum
from typing import Dict, Any, Optional
from types import SimpleNamespace

from google.cloud import vision
from lib.app_config import config

log = logging.getLogger(__name__)


class InitialValidationResult:
    """DTO for the result of the initial validation."""
    def __init__(self, success: bool, face_annotation: Optional[vision.FaceAnnotation] = None, reason: Optional[ValidationReason] = None, details: Optional[str] = None, raw_response_dict: Optional[Dict] = None):
        self.success = success
        self.face_annotation = face_annotation
        self.reason = reason
        self.details = details
        self.raw_response_dict = raw_response_dict


class ValidationReason(str, Enum):
    """Validation status reasons."""
    NO_FACE = "NO_FACE"
    BG = "BG"
    POSE_DIR = "POSE_DIR"
    POSE_CTR = "POSE_CTR"
    EXPR = "EXPR"
    LIGHT = "LIGHT"
    EYES_OBS = "EYES_OBS"
    EYES_CL = "EYES_CL"
    GLS_GLARE = "GLS_GLARE"
    GLS_TINT = "GLS_TINT"
    HD_COVER = "HD_COVER"
    QUAL_BLUR = "QUAL_BLUR"
    QUAL_REDEYE = "QUAL_REDEYE"
    CTRY_DIM = "CTRY_DIM"
    CTRY_SPEC = "CTRY_SPEC"
    ICAO_OTH = "ICAO_OTH"
    INVALID_RESPONSE = "INVALID_RESPONSE"
    PARSE_ERROR = "PARSE_ERROR"
    EMPTY_RESPONSE = "EMPTY_RESPONSE"
    VALIDATION_ERROR = "VALIDATION_ERROR"

    @property
    def description(self) -> str:
        descriptions = {
            ValidationReason.NO_FACE: "No face detected.",
            ValidationReason.BG: "Background invalid.",
            ValidationReason.POSE_DIR: "Head tilt/rotation.",
            ValidationReason.POSE_CTR: "Head not centered.",
            ValidationReason.EXPR: "Expression not neutral.",
            ValidationReason.LIGHT: "Lighting issues.",
            ValidationReason.EYES_OBS: "Eyes obscured.",
            ValidationReason.EYES_CL: "Eyes closed.",
            ValidationReason.GLS_GLARE: "Glasses glare.",
            ValidationReason.GLS_TINT: "Tinted glasses.",
            ValidationReason.HD_COVER: "Head covering.",
            ValidationReason.QUAL_BLUR: "Blurry image.",
            ValidationReason.QUAL_REDEYE: "Red-eye detected.",
            ValidationReason.CTRY_DIM: "Invalid dimensions.",
            ValidationReason.CTRY_SPEC: "Specific rule violation.",
            ValidationReason.ICAO_OTH: "ICAO compliance issue.",
            ValidationReason.INVALID_RESPONSE: "Invalid service response.",
            ValidationReason.PARSE_ERROR: "Response parsing error.",
            ValidationReason.EMPTY_RESPONSE: "Empty response.",
            ValidationReason.VALIDATION_ERROR: "Validation error.",
        }
        return descriptions.get(self, self.value)


class CloudVisionValidator:
    """Validates photos using Cloud Vision API."""
    
    # Validation thresholds
    LANDMARKING_CONFIDENCE_THRESHOLD = 0.6
    BACKGROUND_LIGHT_RGB_THRESHOLD = 200
    BACKGROUND_UNIFORMITY_THRESHOLD = 0.6
    BACKGROUND_SECONDARY_COLOR_THRESHOLD = 0.4
    HEAD_CENTERING_TOLERANCE_RATIO = 0.05
    RED_EYE_RGB_THRESHOLD = 150
    PUPIL_ROI_RADIUS_RATIO = 0.01
    PUPIL_ROI_MIN_RADIUS = 5
    
    def __init__(self):
        self.client = vision.ImageAnnotatorClient()
        self.config = config.icao
        
        self.FAIL_LIKELIHOODS = {
            vision.Likelihood.POSSIBLE,
            vision.Likelihood.LIKELY,
            vision.Likelihood.VERY_LIKELY,
        }
        
    def _encode_image(self, image_bgr: np.ndarray) -> bytes:
        success, buffer = cv2.imencode('.jpg', image_bgr, [cv2.IMWRITE_JPEG_QUALITY, 95])
        if not success:
            raise ValueError("Failed to encode image as JPEG")
        return buffer.tobytes()

    def _validate_dimensions(self, face_data: SimpleNamespace, image_bgr: np.ndarray) -> Optional[ValidationReason]:
        img_h, img_w = image_bgr.shape[:2]
        aspect_ratio = img_w / img_h if img_h > 0 else 0
        target_ratio = self.config.target_aspect_ratio
        
        if abs(aspect_ratio - target_ratio) > self.config.aspect_ratio_tolerance:
            return ValidationReason.CTRY_DIM
        
        if not hasattr(face_data, 'bbox'):
            return None
        
        bbox = face_data.bbox
        head_height_px = bbox[3] - bbox[1]
        head_height_ratio = head_height_px / img_h if img_h > 0 else 0
        
        if head_height_ratio < self.config.min_chin_to_crown_ratio:
            return ValidationReason.CTRY_DIM
        if head_height_ratio > self.config.max_chin_to_crown_ratio:
            return ValidationReason.CTRY_DIM
        
        return None

    def _validate_red_eye(self, image_bgr: np.ndarray, face: vision.FaceAnnotation) -> Optional[ValidationReason]:
        landmarks_by_type = {lm.type_: lm for lm in face.landmarks}
        left_pupil = landmarks_by_type.get(vision.FaceAnnotation.Landmark.Type.LEFT_EYE_PUPIL)
        right_pupil = landmarks_by_type.get(vision.FaceAnnotation.Landmark.Type.RIGHT_EYE_PUPIL)

        if not left_pupil or not right_pupil:
            return None

        pupil_roi_radius = max(self.PUPIL_ROI_MIN_RADIUS, int(image_bgr.shape[0] * self.PUPIL_ROI_RADIUS_RATIO))
        
        for _, pupil_landmark in [("left", left_pupil), ("right", right_pupil)]:
            x, y = int(pupil_landmark.position.x), int(pupil_landmark.position.y)
            roi = image_bgr[max(0, y - pupil_roi_radius):min(image_bgr.shape[0], y + pupil_roi_radius),
                           max(0, x - pupil_roi_radius):min(image_bgr.shape[1], x + pupil_roi_radius)]
            
            if roi.size == 0: continue
            
            b, g, r = cv2.split(roi)
            bg = cv2.add(b, g)
            mask = (r > self.RED_EYE_RGB_THRESHOLD) & (r > bg)
            
            if np.count_nonzero(mask) / (roi.shape[0] * roi.shape[1]) > self.config.red_eye_pixel_percentage_thresh:
                return ValidationReason.QUAL_REDEYE
        
        return None

    def _validate_background_dominant_colors(self, response: vision.AnnotateImageResponse) -> Optional[ValidationReason]:
        if not response.image_properties_annotation or not response.image_properties_annotation.dominant_colors:
            return None
        
        colors = response.image_properties_annotation.dominant_colors.colors
        if not colors: return None
        
        top_color = colors[0]
        rgb = top_color.color
        
        if not (rgb.red > self.BACKGROUND_LIGHT_RGB_THRESHOLD and 
                rgb.green > self.BACKGROUND_LIGHT_RGB_THRESHOLD and 
                rgb.blue > self.BACKGROUND_LIGHT_RGB_THRESHOLD):
            return ValidationReason.BG
        
        if top_color.pixel_fraction <= self.BACKGROUND_UNIFORMITY_THRESHOLD:
            return ValidationReason.BG
        
        if len(colors) > 1 and colors[1].pixel_fraction > self.BACKGROUND_SECONDARY_COLOR_THRESHOLD:
            return ValidationReason.BG
        
        return None

    def _validate_head_centered(self, face_data: SimpleNamespace, image_bgr: np.ndarray) -> Optional[ValidationReason]:
        if not hasattr(face_data, 'bbox'): return None
        
        img_w = image_bgr.shape[1]
        bbox = face_data.bbox
        face_center_x = (bbox[0] + bbox[2]) / 2
        
        if abs(face_center_x - (img_w / 2)) > (img_w * self.HEAD_CENTERING_TOLERANCE_RATIO):
            return ValidationReason.POSE_CTR
        
        return None

    def _validate_eyes_visible(self, face: vision.FaceAnnotation) -> Optional[ValidationReason]:
        if face.landmarking_confidence < self.LANDMARKING_CONFIDENCE_THRESHOLD:
            return ValidationReason.EYES_OBS
        return None

    def _validate_glasses_glare(self, response: vision.AnnotateImageResponse, face: vision.FaceAnnotation) -> Optional[ValidationReason]:
        has_glasses = any("glasses" in l.description.lower() or "eyeglass" in l.description.lower() 
                         for l in (response.label_annotations or []))
        
        if not has_glasses: return None
        
        landmarks = {lm.type_: lm for lm in face.landmarks}
        if vision.FaceAnnotation.Landmark.Type.LEFT_EYE in landmarks and \
           vision.FaceAnnotation.Landmark.Type.RIGHT_EYE in landmarks:
            return None
            
        return ValidationReason.GLS_GLARE

    def _build_response(self, status: str, reason: Optional[ValidationReason] = None, raw_data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        if status == "COMPLIANT":
            return {"success": True, "status": "COMPLIANT", "validation_response": raw_data or {}}
        
        reason_enum = reason or ValidationReason.VALIDATION_ERROR
        return {
            "success": False,
            "status": "ERROR" if status == "ERROR" else "NON_COMPLIANT",
            "status_reason": reason_enum.value,
            "status_reason_description": reason_enum.description,
            "error": reason_enum.description if status == "ERROR" else None,
            "validation_response": raw_data or {}
        }

    def _run_initial_checks(self, response: vision.AnnotateImageResponse, image_bgr: np.ndarray) -> Dict[str, Any]:
        face = response.face_annotations[0]

        if face.blurred_likelihood in self.FAIL_LIKELIHOODS:
            return self._build_response("NON_COMPLIANT", ValidationReason.QUAL_BLUR)
        
        if (reason := self._validate_red_eye(image_bgr, face)):
            return self._build_response("NON_COMPLIANT", reason)
        
        if face.under_exposed_likelihood in self.FAIL_LIKELIHOODS:
            return self._build_response("NON_COMPLIANT", ValidationReason.LIGHT)
        
        if abs(face.roll_angle) > self.config.max_abs_roll or \
           abs(face.pan_angle) > self.config.max_abs_yaw or \
           abs(face.tilt_angle) > self.config.max_abs_pitch:
            return self._build_response("NON_COMPLIANT", ValidationReason.POSE_DIR)

        if any(l in self.FAIL_LIKELIHOODS for l in [face.joy_likelihood, face.sorrow_likelihood, face.anger_likelihood]):
            return self._build_response("NON_COMPLIANT", ValidationReason.EXPR)
        
        if (reason := self._validate_eyes_visible(face)):
            return self._build_response("NON_COMPLIANT", reason)
        
        if (reason := self._validate_glasses_glare(response, face)):
            return self._build_response("NON_COMPLIANT", reason)
        
        if any("sunglasses" in l.description.lower() for l in (response.label_annotations or [])):
            return self._build_response("NON_COMPLIANT", ValidationReason.GLS_TINT)
            
        if face.headwear_likelihood in self.FAIL_LIKELIHOODS:
             return self._build_response("NON_COMPLIANT", ValidationReason.HD_COVER)

        return self._build_response("COMPLIANT")

    def validate_initial(self, image_bgr: np.ndarray) -> InitialValidationResult:
        try:
            image_bytes = self._encode_image(image_bgr)
            image = vision.Image(content=image_bytes)
            
            features = [
                vision.Feature(type_=vision.Feature.Type.FACE_DETECTION, max_results=1),
                vision.Feature(type_=vision.Feature.Type.LABEL_DETECTION, max_results=5),
                vision.Feature(type_=vision.Feature.Type.IMAGE_PROPERTIES),
                vision.Feature(type_=vision.Feature.Type.OBJECT_LOCALIZATION),
            ]
            
            response = self.client.annotate_image(request=vision.AnnotateImageRequest(image=image, features=features))
            raw_response = vision.AnnotateImageResponse.to_dict(response)
            
            if response.error.message:
                return InitialValidationResult(success=False, reason=ValidationReason.VALIDATION_ERROR, details=response.error.message, raw_response_dict=raw_response)
            
            if not response.face_annotations:
                return InitialValidationResult(success=False, reason=ValidationReason.NO_FACE, raw_response_dict=raw_response)

            if len(response.face_annotations) > 1:
                return InitialValidationResult(success=False, reason=ValidationReason.ICAO_OTH, details="Multiple faces", raw_response_dict=raw_response)

            result = self._run_initial_checks(response, image_bgr)
            
            if not result["success"]:
                return InitialValidationResult(
                    success=False,
                    reason=ValidationReason(result.get("status_reason")),
                    details=result.get("validation_response", {}).get("details"),
                    raw_response_dict=raw_response
                )

            return InitialValidationResult(success=True, face_annotation=response.face_annotations[0], raw_response_dict=raw_response)

        except Exception as e:
            log.error(f"Validation error: {e}", exc_info=True)
            return InitialValidationResult(success=False, reason=ValidationReason.VALIDATION_ERROR, details=str(e))

    def validate_final_geometry(self, processed_bgr: np.ndarray, final_face_data: SimpleNamespace) -> Dict[str, Any]:
        if (reason := self._validate_dimensions(final_face_data, processed_bgr)):
            return self._build_response("NON_COMPLIANT", reason)
        
        if (reason := self._validate_head_centered(final_face_data, processed_bgr)):
            return self._build_response("NON_COMPLIANT", reason)
            
        return self._build_response("COMPLIANT")

File: server/cloud-vision-validator.ts
/**
 * Google Cloud Vision API-based photo validation module.
 * Handles face detection, pose analysis, and ICAO compliance checks.
 */

import { ImageAnnotatorClient, protos } from '@google-cloud/vision'
import { getVercelOidcToken } from '@vercel/functions/oidc'
import { ExternalAccountClient } from 'google-auth-library'
import {
  ICAOConfig,
  ValidationThresholds,
  ValidationReason,
  ValidationReasonDescriptions,
  FailLikelihoods,
  LandmarkType,
  type ValidationReasonType,
  type FaceData,
  type InitialValidationResult,
} from './validation-constants.js'

type IFaceAnnotation = protos.google.cloud.vision.v1.IFaceAnnotation
type IAnnotateImageResponse = protos.google.cloud.vision.v1.IAnnotateImageResponse
type ILandmark = protos.google.cloud.vision.v1.FaceAnnotation.ILandmark

let visionClient: ImageAnnotatorClient

function getVisionClient(): ImageAnnotatorClient {
  if (visionClient) return visionClient

  if (process.env.USE_LOCAL_STORAGE !== 'true') {
    const GCP_PROJECT_ID = process.env.GCP_PROJECT_ID
    const GCP_PROJECT_NUMBER = process.env.GCP_PROJECT_NUMBER
    const GCP_SERVICE_ACCOUNT_EMAIL = process.env.GCP_SERVICE_ACCOUNT_EMAIL
    const GCP_WORKLOAD_IDENTITY_POOL_ID = process.env.GCP_WORKLOAD_IDENTITY_POOL_ID
    const GCP_WORKLOAD_IDENTITY_POOL_PROVIDER_ID = process.env.GCP_WORKLOAD_IDENTITY_POOL_PROVIDER_ID

    const authClient = ExternalAccountClient.fromJSON({
      type: 'external_account',
      audience: `//iam.googleapis.com/projects/${GCP_PROJECT_NUMBER}/locations/global/workloadIdentityPools/${GCP_WORKLOAD_IDENTITY_POOL_ID}/providers/${GCP_WORKLOAD_IDENTITY_POOL_PROVIDER_ID}`,
      subject_token_type: 'urn:ietf:params:oauth:token-type:jwt',
      token_url: 'https://sts.googleapis.com/v1/token',
      service_account_impersonation_url: `https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/${GCP_SERVICE_ACCOUNT_EMAIL}:generateAccessToken`,
      subject_token_supplier: {
        getSubjectToken: getVercelOidcToken,
      },
    })

    if (!authClient) {
      throw new Error('Failed to initialize External Account Client for Vision API')
    }

    visionClient = new ImageAnnotatorClient({
      authClient: authClient as never,
      projectId: GCP_PROJECT_ID,
    })
  } else {
    // Local development - use default credentials
    visionClient = new ImageAnnotatorClient()
  }

  return visionClient
}

// Helper to get landmarks by type
function getLandmarksByType(face: IFaceAnnotation): Map<string, ILandmark> {
  const map = new Map<string, ILandmark>()
  for (const lm of face.landmarks || []) {
    if (lm.type) {
      map.set(lm.type.toString(), lm)
    }
  }
  return map
}

// Validate eyes are visible and open
function validateEyesVisible(face: IFaceAnnotation): ValidationReasonType | null {
  const landmarks = getLandmarksByType(face)
  const leftPupil = landmarks.get(LandmarkType.LEFT_EYE_PUPIL)
  const rightPupil = landmarks.get(LandmarkType.RIGHT_EYE_PUPIL)

  if (leftPupil && rightPupil) {
    return null
  }

  // Fallback: check eye opening using boundary landmarks
  const leftTop = landmarks.get(LandmarkType.LEFT_EYE_TOP_BOUNDARY)
  const leftBottom = landmarks.get(LandmarkType.LEFT_EYE_BOTTOM_BOUNDARY)
  const rightTop = landmarks.get(LandmarkType.RIGHT_EYE_TOP_BOUNDARY)
  const rightBottom = landmarks.get(LandmarkType.RIGHT_EYE_BOTTOM_BOUNDARY)


  // Check if we have boundary landmarks for at least one eye
  if (!(leftTop && leftBottom) && !(rightTop && rightBottom)) {
    return ValidationReason.EYES_CL
  }

  // Check left eye opening
  if (leftTop && leftBottom && leftTop.position?.y && leftBottom.position?.y) {
    const leftOpening = Math.abs(leftTop.position.y - leftBottom.position.y)
    if (leftOpening < ValidationThresholds.minEyeOpeningPixels) {
      return ValidationReason.EYES_CL
    }
  }

  // Check right eye opening
  if (rightTop && rightBottom && rightTop.position?.y && rightBottom.position?.y) {
    const rightOpening = Math.abs(rightTop.position.y - rightBottom.position.y)
    if (rightOpening < ValidationThresholds.minEyeOpeningPixels) {
      return ValidationReason.EYES_CL
    }
  }

  return null
}

// Validate glasses glare
function validateGlassesGlare(
  response: IAnnotateImageResponse,
  face: IFaceAnnotation
): ValidationReasonType | null {
  const hasGlasses = (response.labelAnnotations || []).some(
    (l) =>
      l.description?.toLowerCase().includes('glasses') ||
      l.description?.toLowerCase().includes('eyeglass')
  )

  if (!hasGlasses) return null

  const landmarks = getLandmarksByType(face)
  if (landmarks.has(LandmarkType.LEFT_EYE) && landmarks.has(LandmarkType.RIGHT_EYE)) {
    return null
  }

  return ValidationReason.GLS_GLARE
}

// Validate cheek centers


// Validate head pose
function validateHeadPose(face: IFaceAnnotation): ValidationReasonType | null {
  const rollAngle = face.rollAngle ?? 0
  const panAngle = face.panAngle ?? 0
  const tiltAngle = face.tiltAngle ?? 0

  if (
    Math.abs(rollAngle) > ICAOConfig.maxAbsRoll ||
    Math.abs(panAngle) > ICAOConfig.maxAbsYaw ||
    Math.abs(tiltAngle) > ICAOConfig.maxAbsPitch
  ) {
    return ValidationReason.POSE_DIR
  }

  return null
}

// Validate expression is neutral
function validateExpression(face: IFaceAnnotation): ValidationReasonType | null {
  const likelihoods = [
    face.joyLikelihood,
    face.sorrowLikelihood,
    face.angerLikelihood,
    face.surpriseLikelihood,
  ]

  if (likelihoods.some((l) => l && FailLikelihoods.has(l.toString()))) {
    return ValidationReason.EXPR
  }

  return null
}

// Validate headwear
function validateHeadwear(face: IFaceAnnotation): ValidationReasonType | null {
  if (face.headwearLikelihood && FailLikelihoods.has(face.headwearLikelihood.toString())) {
    return ValidationReason.HD_COVER
  }
  return null
}

// Validate sunglasses
function validateSunglasses(response: IAnnotateImageResponse): ValidationReasonType | null {
  const hasSunglasses = (response.labelAnnotations || []).some((l) =>
    l.description?.toLowerCase().includes('sunglasses')
  )

  if (hasSunglasses) {
    return ValidationReason.GLS_TINT
  }

  return null
}

// Validate blur
function validateBlur(face: IFaceAnnotation): ValidationReasonType | null {
  if (face.blurredLikelihood && FailLikelihoods.has(face.blurredLikelihood.toString())) {
    return ValidationReason.QUAL_BLUR
  }
  return null
}

// Validate under-exposure
function validateUnderExposure(face: IFaceAnnotation): ValidationReasonType | null {
  if (face.underExposedLikelihood && FailLikelihoods.has(face.underExposedLikelihood.toString())) {
    return ValidationReason.LIGHT
  }
  return null
}

// Validate confidence levels
function validateConfidence(face: IFaceAnnotation): ValidationReasonType | null {
  if ((face.landmarkingConfidence ?? 0) < ValidationThresholds.landmarkingConfidence) {
    return ValidationReason.VALIDATION_ERROR
  }
  if ((face.detectionConfidence ?? 0) < ValidationThresholds.detectionConfidence) {
    return ValidationReason.VALIDATION_ERROR
  }
  return null
}

// Run all initial validation checks
function runInitialChecks(
  response: IAnnotateImageResponse,
  face: IFaceAnnotation
): { success: boolean; reason?: ValidationReasonType } {
  // Confidence checks
  let reason = validateConfidence(face)
  if (reason) return { success: false, reason }

  // Blur check
  reason = validateBlur(face)
  if (reason) return { success: false, reason }

  // Eyes visibility
  reason = validateEyesVisible(face)
  if (reason) return { success: false, reason }

  // Under-exposure
  reason = validateUnderExposure(face)
  if (reason) return { success: false, reason }

  // Head pose
  reason = validateHeadPose(face)
  if (reason) return { success: false, reason }

  // Expression
  reason = validateExpression(face)
  if (reason) return { success: false, reason }

  // Glasses glare
  reason = validateGlassesGlare(response, face)
  if (reason) return { success: false, reason }

  // Sunglasses
  reason = validateSunglasses(response)
  if (reason) return { success: false, reason }

  // Headwear
  reason = validateHeadwear(face)
  if (reason) return { success: false, reason }

  return { success: true }
}

/**
 * Performs initial validation using Cloud Vision API.
 * Checks for face detection, blur, pose, expression, etc.
 */
export async function validateInitial(imageBuffer: Buffer): Promise<InitialValidationResult> {
  try {
    const client = getVisionClient()

    const [result] = await client.annotateImage({
      image: { content: imageBuffer.toString('base64') },
      features: [
        { type: 'FACE_DETECTION', maxResults: 1 },
        { type: 'LABEL_DETECTION', maxResults: 5 },
        { type: 'IMAGE_PROPERTIES' },
      ],
    })

    if (result.error?.message) {
      return {
        success: false,
        reason: ValidationReason.VALIDATION_ERROR,
        details: result.error.message,
      }
    }

    const faceAnnotations = result.faceAnnotations || []

    if (faceAnnotations.length === 0) {
      return {
        success: false,
        reason: ValidationReason.NO_FACE,
      }
    }

    if (faceAnnotations.length > 1) {
      return {
        success: false,
        reason: ValidationReason.ICAO_OTH,
        details: 'Multiple faces detected',
      }
    }

    const face = faceAnnotations[0]
    const checkResult = runInitialChecks(result, face)

    if (!checkResult.success) {
      return {
        success: false,
        reason: checkResult.reason,
        details: checkResult.reason
          ? ValidationReasonDescriptions[checkResult.reason]
          : undefined,
      }
    }

    return {
      success: true,
      faceAnnotation: face,
    }
  } catch (error) {
    console.error('Vision API validation error:', error)
    return {
      success: false,
      reason: ValidationReason.VALIDATION_ERROR,
      details: error instanceof Error ? error.message : 'Unknown error',
    }
  }
}

/**
 * Validates final geometry of processed image.
 * Checks dimensions and head centering.
 */
export function validateFinalGeometry(
  imageWidth: number,
  imageHeight: number,
  faceData: FaceData
): { success: boolean; reason?: ValidationReasonType } {
  // Check aspect ratio
  const aspectRatio = imageWidth / imageHeight
  if (Math.abs(aspectRatio - ICAOConfig.targetAspectRatio) > ICAOConfig.aspectRatioTolerance) {
    return { success: false, reason: ValidationReason.CTRY_DIM }
  }

  // Check head height ratio
  const headHeightPx = faceData.bbox[3] - faceData.bbox[1]
  const headHeightRatio = headHeightPx / imageHeight

  if (headHeightRatio < ICAOConfig.minChinToCrownRatio) {
    return { success: false, reason: ValidationReason.CTRY_DIM }
  }
  if (headHeightRatio > ICAOConfig.maxChinToCrownRatio) {
    return { success: false, reason: ValidationReason.CTRY_DIM }
  }

  // Check head centering
  const faceCenterX = (faceData.bbox[0] + faceData.bbox[2]) / 2
  const imageCenterX = imageWidth / 2
  const tolerance = imageWidth * ValidationThresholds.headCenteringToleranceRatio

  if (Math.abs(faceCenterX - imageCenterX) > tolerance) {
    return { success: false, reason: ValidationReason.POSE_CTR }
  }

  return { success: true }
}

/**
 * Extracts face details from Cloud Vision face annotation for preprocessing.
 */
export function extractFaceDetails(
  faceAnnotation: IFaceAnnotation
): { faceDetails: FaceData; error?: string } | { faceDetails?: undefined; error: string } {
  const fullPoly = faceAnnotation.boundingPoly
  if (!fullPoly?.vertices?.length) {
    return { error: 'Full bounding polygon not found.' }
  }

  const trueCrownY = Math.min(...fullPoly.vertices.map((v) => v.y ?? 0))
  const fdPoly = faceAnnotation.fdBoundingPoly

  if (!fdPoly?.vertices?.length) {
    return { error: 'FD bounding polygon not found.' }
  }

  const fdYCoords = fdPoly.vertices.map((v) => v.y ?? 0)
  const fdXCoords = fdPoly.vertices.map((v) => v.x ?? 0)

  const chinY = Math.max(...fdYCoords)
  const bbox: [number, number, number, number] = [
    Math.min(...fdXCoords),
    trueCrownY,
    Math.max(...fdXCoords),
    chinY,
  ]

  const landmarks =
    faceAnnotation.landmarks?.map((lm) => ({
      x: lm.position?.x ?? 0,
      y: lm.position?.y ?? 0,
    })) || []

  return {
    faceDetails: {
      bbox,
      landmarks,
      crownY: trueCrownY,
      chinY,
    },
  }
}

